

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Pipeliner 1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html#document-index" class="icon icon-home"> Pipeliner
          

          
          </a>

          

          

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
            
              <!-- Local TOC -->
              <div class="local-toc"><ul>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-requirements">Requirements</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#testing-nextflow">Testing Nextflow</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#installing-anaconda">Installing Anaconda</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#pre-packaged-conda-environment">Pre-Packaged Conda Environment</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#anaconda-cloud">Anaconda Cloud</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#yaml-file">Yaml File</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#setting-up-pipeliner">Setting up Pipeliner</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-framework-structure">Framework Structure</a><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-pipeline-structure">Pipeline Structure</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#input">Input</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#configuration-file">Configuration File</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#file-paths">File Paths</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#executor-and-compute-resources">Executor and Compute Resources</a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#pipeline-options-and-parameters">Pipeline Options and Parameters</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#pipeline-script">Pipeline Script</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#template-processes">Template Processes</a></li>
<li class="toctree-l2"><a class="reference internal" href="index.html#output">Output</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-existing-pipelines">Existing Pipelines</a><ul>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-rna-seq">RNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#check-reads-check-reads">Check Reads <code class="code docutils literal notranslate"><span class="pre">check_reads</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#genome-indexing-hisat-indexing-star-indexing">Genome Indexing <code class="code docutils literal notranslate"><span class="pre">hisat_indexing/star_indexing</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#pre-quality-check-pre-fastqc">Pre-Quality Check <code class="code docutils literal notranslate"><span class="pre">pre_fastqc</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#pre-multiqc-pre-multiqc">Pre-MultiQC <code class="code docutils literal notranslate"><span class="pre">pre_multiqc</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#read-trimming-trim-galore">Read Trimming <code class="code docutils literal notranslate"><span class="pre">trim_galore</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#read-mapping-hisat-mapping-star-mapping">Read Mapping <code class="code docutils literal notranslate"><span class="pre">hisat_mapping/star_mapping</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#reformat-reference-gtftobed">Reformat Reference <code class="code docutils literal notranslate"><span class="pre">gtftobed</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#mapping-quality-rseqc">Mapping Quality <code class="code docutils literal notranslate"><span class="pre">rseqc</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#quantification-counting">Quantification <code class="code docutils literal notranslate"><span class="pre">counting</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#expression-matrix-expression-matrix">Expression Matrix <code class="code docutils literal notranslate"><span class="pre">expression_matrix</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#expression-features-expression-features">Expression Features <code class="code docutils literal notranslate"><span class="pre">expression_features</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#expression-set-expression-set">Expression Set <code class="code docutils literal notranslate"><span class="pre">expression_set</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#summary-report-multiqc">Summary Report <code class="code docutils literal notranslate"><span class="pre">multiqc</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-scrna-seq">scRNA-seq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#check-reads-check-reads">Check Reads <code class="code docutils literal notranslate"><span class="pre">check_reads</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#genome-indexing-hisat-indexing-star-indexing">Genome Indexing <code class="code docutils literal notranslate"><span class="pre">hisat_indexing/star_indexing</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#quality-check-fastqc">Quality Check <code class="code docutils literal notranslate"><span class="pre">fastqc</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#whitelist-whitelist">Whitelist <code class="code docutils literal notranslate"><span class="pre">whitelist</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#extract-extract">Extract <code class="code docutils literal notranslate"><span class="pre">extract</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#read-mapping-hisat-mapping-star-mapping">Read Mapping <code class="code docutils literal notranslate"><span class="pre">hisat_mapping/star_mapping</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#reformat-reference-gtftobed">Reformat Reference <code class="code docutils literal notranslate"><span class="pre">gtftobed</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#mapping-quality-rseqc">Mapping Quality <code class="code docutils literal notranslate"><span class="pre">rseqc</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#quantification-counting">Quantification <code class="code docutils literal notranslate"><span class="pre">counting</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#summary-report-multiqc">Summary Report <code class="code docutils literal notranslate"><span class="pre">multiqc</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="index.html#document-dge">DGE</a><ul>
<li class="toctree-l3"><a class="reference internal" href="index.html#quantification-counting">Quantification <code class="code docutils literal notranslate"><span class="pre">counting</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#expression-matrix-expression-matrix">Expression Matrix <code class="code docutils literal notranslate"><span class="pre">expression_matrix</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#sample-renaming-rename-samples">Sample Renaming <code class="code docutils literal notranslate"><span class="pre">rename_samples</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="index.html#summary-report-multiqc">Summary Report <code class="code docutils literal notranslate"><span class="pre">multiqc</span></code></a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="index.html#document-extending-pipelines">Extending Pipelines</a><ul class="simple">
</ul>
</li>
</ul>
</div>
            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html#document-index">Pipeliner</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html#document-index">Docs</a> &raquo;</li>
        
      <li>Pipeliner 1 documentation</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pipeliner-documenation">
<h1>Pipeliner Documenation<a class="headerlink" href="#pipeliner-documenation" title="Permalink to this headline">¶</a></h1>
<div class="toctree-wrapper compound">
<span id="document-requirements"></span><div class="section" id="requirements">
<span id="requirements-page"></span><h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
<p>The <cite>Pipeliner</cite> framework requires <cite>Nextflow</cite> and <cite>Anaconda</cite>. <cite>Nextflow</cite> can be used on any POSIX compatible system (Linux, OS X, etc). It requires BASH and <a class="reference external" href="http://www.oracle.com/technetwork/java/javase/downloads/index.html">Java 8 (or higher)</a> to be installed. Third-party software tools used by individual pipelines will be installed and managed through a Conda virtual environment.</p>
<div class="section" id="testing-nextflow">
<h3>Testing Nextflow<a class="headerlink" href="#testing-nextflow" title="Permalink to this headline">¶</a></h3>
<p>Before continuuing, test to make sure your environment is compatible with a Nextflow executable.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">You will download another one later when you clone the repository</p>
</div>
<p>Make sure your Java installation is version 8 or higher:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">java</span> <span class="o">-</span><span class="n">version</span>
</pre></div>
</div>
<p>Create a new directory and install/test <cite>Nextflow</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">nf</span><span class="o">-</span><span class="n">test</span>
<span class="n">cd</span> <span class="n">nf</span><span class="o">-</span><span class="n">test</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">s</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">get</span><span class="o">.</span><span class="n">nextflow</span><span class="o">.</span><span class="n">io</span> <span class="o">|</span> <span class="n">bash</span>
<span class="o">./</span><span class="n">nextflow</span> <span class="n">run</span> <span class="n">hello</span>
</pre></div>
</div>
<p>Output:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>N E X T F L O W  ~  version 0.31.0
Launching `nextflow-io/hello` [sad_curran] - revision: d4c9ea84de [master]
[warm up] executor &gt; local
[4d/479eec] Submitted process &gt; sayHello (4)
[a8/4bc038] Submitted process &gt; sayHello (2)
[17/5be64e] Submitted process &gt; sayHello (3)
[ee/0d879f] Submitted process &gt; sayHello (1)
Hola world!
Ciao world!
Hello world!
Bonjour world!
</pre></div>
</div>
</div>
<div class="section" id="installing-anaconda">
<h3>Installing Anaconda<a class="headerlink" href="#installing-anaconda" title="Permalink to this headline">¶</a></h3>
<p>Pipeliner uses virtual environments managed by <cite>Conda</cite>, which is available through <a class="reference external" href="https://www.continuum.io/downloads">Anaconda</a>. Download the distribution pre-packaged with Python 2.7.</p>
<p>Make sure conda is installed and updated:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="o">--</span><span class="n">version</span>
<span class="n">conda</span> <span class="n">update</span> <span class="n">conda</span>
</pre></div>
</div>
<div class="admonition tip">
<p class="first admonition-title">Tip</p>
<p class="last">If this is your first time working with <cite>Conda</cite>, you may need to edit your configuration paths to ensure <cite>Anaconda</cite> is invoked when calling <code class="docutils literal notranslate"><span class="pre">conda</span></code></p>
</div>
</div>
<div class="section" id="pre-packaged-conda-environment">
<h3>Pre-Packaged Conda Environment<a class="headerlink" href="#pre-packaged-conda-environment" title="Permalink to this headline">¶</a></h3>
<div class="section" id="anaconda-cloud">
<h4>Anaconda Cloud<a class="headerlink" href="#anaconda-cloud" title="Permalink to this headline">¶</a></h4>
<p><cite>Environment for both Linux and OS X</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="n">pipeliner</span> <span class="n">python</span><span class="o">=</span><span class="mf">2.7</span>
<span class="n">source</span> <span class="n">activate</span> <span class="n">pipeliner</span>
<span class="n">conda</span> <span class="n">install</span> <span class="n">pipeliner</span> <span class="o">--</span><span class="n">channel</span> <span class="s2">&quot;pipeliner&quot;</span> <span class="n">trim</span><span class="o">-</span><span class="n">galore</span> <span class="n">fastqc</span> <span class="n">star</span> <span class="n">multiqc</span> <span class="n">samtools</span> <span class="n">rseqc</span> <span class="n">stringtie</span> <span class="n">hisat2</span> <span class="n">htseq</span> <span class="n">subread</span> <span class="n">numpy</span> <span class="n">pandas</span> <span class="n">bioconductor</span><span class="o">-</span><span class="n">biobase</span>
</pre></div>
</div>
</div>
<div class="section" id="yaml-file">
<h4>Yaml File<a class="headerlink" href="#yaml-file" title="Permalink to this headline">¶</a></h4>
<p><cite>Environment for Linux</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-</span><span class="n">f</span> <span class="n">pipeliner</span><span class="o">/</span><span class="n">envs</span><span class="o">/</span><span class="n">linux_env</span><span class="o">.</span><span class="n">yml</span>
</pre></div>
</div>
<p><cite>Environment for OS X</cite>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-</span><span class="n">f</span> <span class="n">pipeliner</span><span class="o">//</span><span class="n">envs</span><span class="o">/</span><span class="n">osx_env</span><span class="o">.</span><span class="n">yml</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="setting-up-pipeliner">
<h3>Setting up Pipeliner<a class="headerlink" href="#setting-up-pipeliner" title="Permalink to this headline">¶</a></h3>
<p>With all prerequisites, one can quickly setup Pipeliner by cloning the repository, configuring local paths to toy datasets, activating the conda environment, and downloading the Nextflow executable:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Activate conda environment</span>
<span class="n">source</span> <span class="n">activate</span> <span class="n">pipeliner</span>

<span class="c1"># Clone Pipeliner</span>
<span class="n">git</span> <span class="n">clone</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">montilab</span><span class="o">/</span><span class="n">pipeliner</span>

<span class="c1"># Configure local paths to toy datasets</span>
<span class="n">python</span> <span class="n">pipeliner</span><span class="o">/</span><span class="n">scripts</span><span class="o">/</span><span class="n">paths</span><span class="o">.</span><span class="n">py</span>

<span class="c1"># Move to pipelines directory</span>
<span class="n">cd</span> <span class="n">pipeliner</span><span class="o">/</span><span class="n">pipelines</span>

<span class="c1"># Download nextflow executable</span>
<span class="n">curl</span> <span class="o">-</span><span class="n">s</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">get</span><span class="o">.</span><span class="n">nextflow</span><span class="o">.</span><span class="n">io</span> <span class="o">|</span> <span class="n">bash</span>

<span class="c1"># Run RNA-seq pipeline with toy data</span>
<span class="o">./</span><span class="n">nextflow</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">nf</span> <span class="o">-</span><span class="n">c</span> <span class="n">rnaseq</span><span class="o">.</span><span class="n">config</span>
</pre></div>
</div>
<p>The output should look like this:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>Launching `rnaseq.nf` [distraught_hugle] - revision: 0e9a7a8940
 P I P E L I N E R  ~  v2.4
====================================
Reads          : /Users/anthonyfederico/pipeliner/RNA-seq/ggal_data/ggal_reads.csv
Reference      : /Users/anthonyfederico/pipeliner/RNA-seq/ggal_data/genome_reference.fa
Annotation     : /Users/anthonyfederico/pipeliner/RNA-seq/ggal_data/genome_annotation.gtf
Input Dir      : /Users/anthonyfederico/pipeliner/RNA-seq/ggal_data
Output Dir     : /Users/anthonyfederico/pipeliner/RNA-seq/ggal_results
====================================
Read Type      : paired-end
Aligner        : hisat
Quantifier     : htseq
Save Reference : true
Save Temporary : true
====================================
Current user  : anthonyfederico
Current home  : /Users/anthonyfederico
Current path  : /Users/anthonyfederico/pipeliner/RNA-seq
====================================
[warm up] executor &gt; local
[32/b1db1d] Submitted process &gt; hisat_indexing (genome_reference.fa)
[3b/d93c6d] Submitted process &gt; trim_galore (ggal_alpha)
[9c/3fa50b] Submitted process &gt; trim_galore (ggal_theta)
[62/25fce0] Submitted process &gt; trim_galore (ggal_gamma)
[96/2ffc07] Submitted process &gt; fastqc (ggal_alpha)
[31/89cddb] Submitted process &gt; fastqc (ggal_theta)
[57/b2fdf0] Submitted process &gt; fastqc (ggal_gamma)
[66/ccc9db] Submitted process &gt; hisat_mapping (ggal_alpha)
[28/69fff5] Submitted process &gt; hisat_mapping (ggal_theta)
[5c/5ed2b6] Submitted process &gt; hisat_mapping (ggal_gamma)
[bc/6f490c] Submitted process &gt; rseqc (ggal_alpha)
[71/80aa9e] Submitted process &gt; rseqc (ggal_theta)
[17/ca0d9f] Submitted process &gt; rseqc (ggal_gamma)
[d7/7d391b] Submitted process &gt; htseq (ggal_alpha)
[df/936854] Submitted process &gt; htseq (ggal_theta)
[11/143c2c] Submitted process &gt; htseq (ggal_gamma)
[1f/3af548] Submitted process &gt; multiqc
Success: Pipeline Completed!
</pre></div>
</div>
</div>
</div>
<span id="document-framework-structure"></span><div class="section" id="framework-structure">
<span id="framework-structure-page"></span><h2>Framework Structure<a class="headerlink" href="#framework-structure" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
<p>Pipeline is a framework with various moving parts to support the development of multiple sequencing pipelines. The following is a simplified example of its directory structure:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>/pipeliner
 ├── /docs
 ├── /envs
 ├── /scripts
 ├── /tests
 └── /pipelines
      ├── /configs
      ├── /scripts
      ├── /templates
      ├── /toy_data
      ├── /rnaseq.nf
      └── /rnaseq.config
</pre></div>
</div>
<dl class="docutils">
<dt><strong>docs</strong></dt>
<dd>Markdown (MD) and Restructured Text (reST) documentaion files associated with Pipeliner and existing pipelines</dd>
<dt><strong>envs</strong></dt>
<dd>Files (YAML) and scripts requires to reproduce Conda environments</dd>
<dt><strong>scripts</strong></dt>
<dd>Various helper scripts for framework setup and maintainence</dd>
<dt><strong>tests</strong></dt>
<dd>Testing module (Python) for multi-pipeline automatic test execution and reporting</dd>
<dt><strong>pipelines/configs</strong></dt>
<dd>Base config files inherited by pipeline configurations</dd>
<dt><strong>pipelines/scripts</strong></dt>
<dd>Various helper scripts for pipeline processes</dd>
<dt><strong>pipelines/templates</strong></dt>
<dd>Template processes inherited by pipeline workflows</dd>
<dt><strong>pipelines/toy_data</strong></dt>
<dd>Small datasets for rapid development and testing of pipelines</dd>
<dt><strong>pipelines/rnaseq.nf</strong></dt>
<dd>Nextflow script for the RNA-seq pipeline</dd>
<dt><strong>pipelines/rnaseq.config</strong></dt>
<dd>Configuration file for the RNA-seq pipeline</dd>
</dl>
</div>
<span id="document-pipeline-structure"></span><div class="section" id="pipeline-structure">
<span id="pipeline-structure-page"></span><h2>Pipeline Structure<a class="headerlink" href="#pipeline-structure" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<p>The file paths for all data fed to a pipeline are specified in the configuration file. To ease the development process, Pipeline provides toy datasets for each of the pipelines. This example will cover the RNA-seq pipeline.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Data for this pipeline is located in <strong>pipelines/toy_data/rna-seq</strong></p>
</div>
<dl class="docutils">
<dt>Users must provide the following files:</dt>
<dd><ul class="first last simple">
<li>Sequencing files or alignment files</li>
<li>Comma-delimited file containing file paths to reads/bams</li>
<li>Genome reference file</li>
<li>Genome annotation file</li>
</ul>
</dd>
</dl>
</div>
<div class="section" id="configuration-file">
<h3>Configuration File<a class="headerlink" href="#configuration-file" title="Permalink to this headline">¶</a></h3>
<p>The configuration file is where all file paths are specified and pipeline processes are paramaterized. The configuration can be broken into three sections, including file paths, executor and compute resources, and pipeline options and parameters.</p>
<div class="section" id="file-paths">
<h4>File Paths<a class="headerlink" href="#file-paths" title="Permalink to this headline">¶</a></h4>
<p>The configuration file specifies where to find all of the input data. Additionally, it provide a path to an output directory where the pipeline will output results. The following is a typical example for the RNA-seq configuration file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">indir</span>  <span class="o">=</span> <span class="s2">&quot;/Users/anthonyfederico/pipeliner/pipelines/toy_data/rna-seq&quot;</span>
<span class="n">outdir</span> <span class="o">=</span> <span class="s2">&quot;/Users/anthonyfederico/pipeliner/pipelines/rna-seq-results&quot;</span>
<span class="n">fasta</span>  <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{params.indir}</span><span class="s2">/genome_reference.fa&quot;</span>
<span class="n">gtf</span>    <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{params.indir}</span><span class="s2">/genome_annotation.gtf&quot;</span>
<span class="n">reads</span>  <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{params.indir}</span><span class="s2">/ggal_reads.csv&quot;</span>
</pre></div>
</div>
</div>
<div class="section" id="executor-and-compute-resources">
<h4>Executor and Compute Resources<a class="headerlink" href="#executor-and-compute-resources" title="Permalink to this headline">¶</a></h4>
<p>An abstraction layer between Nextflow and Pipeliner logic enables platform independence and seamless compatibility with high performance computing&nbsp;executors. This allows users to execute pipelines on their local machine or through a computing cluster by simply specifying in the configuration file.</p>
<p>Pipeliner provides two base configuration files that can be inherited depending if a pipeline is being executing using local resources or a Sun Grid Engine (SGE) queuing system.</p>
<p>If the latter is chosen, pipeline processes will be automatically parallelized. Additionally, each individual process can be allocated specific computing resource instructions when nodes are requested.</p>
<p><em>Local config example</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">process</span> <span class="p">{</span>
  <span class="n">executor</span> <span class="o">=</span> <span class="s1">&#39;local&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
<p><em>Cloud computing config example</em>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>process {
  executor = &#39;sge&#39;
  scratch = true

  $trim_galore.clusterOptions            = &quot;-P montilab -l h_rt=24:00:00 -pe omp 8&quot;
  $star_mapping.clusterOptions           = &quot;-P montilab -l h_rt=24:00:00 -l mem_total=94G -pe omp 16&quot;
  $counting.clusterOptions               = &quot;-P montilab -l h_rt=24:00:00 -pe omp 8&quot;
  $expression_matrix.clusterOptions      = &quot;-P montilab -l h_rt=24:00:00 -pe omp 8&quot;
  $multiqc.clusterOptions                = &quot;-P montilab -l h_rt=24:00:00 -pe omp 8&quot;
}
</pre></div>
</div>
</div>
<div class="section" id="pipeline-options-and-parameters">
<h4>Pipeline Options and Parameters<a class="headerlink" href="#pipeline-options-and-parameters" title="Permalink to this headline">¶</a></h4>
<p>The rest of the configuration file is dedicated to the different pipeline options and process parameters that can be specified. Some import examples include the following:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># General pipeline parameters</span>
<span class="n">aligner</span>      <span class="o">=</span> <span class="s2">&quot;hisat&quot;</span>
<span class="n">quantifier</span>   <span class="o">=</span> <span class="s2">&quot;htseq&quot;</span>

<span class="c1"># Process-specific parameters</span>
<span class="n">htseq</span><span class="o">.</span><span class="n">type</span>   <span class="o">=</span> <span class="s2">&quot;exon&quot;</span>
<span class="n">htseq</span><span class="o">.</span><span class="n">mode</span>   <span class="o">=</span> <span class="s2">&quot;union&quot;</span>
<span class="n">htseq</span><span class="o">.</span><span class="n">idattr</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span>
<span class="n">htseq</span><span class="o">.</span><span class="n">order</span>  <span class="o">=</span> <span class="s2">&quot;pos&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="section" id="pipeline-script">
<h3>Pipeline Script<a class="headerlink" href="#pipeline-script" title="Permalink to this headline">¶</a></h3>
</div>
<div class="section" id="template-processes">
<h3>Template Processes<a class="headerlink" href="#template-processes" title="Permalink to this headline">¶</a></h3>
<p>Pipelines written in Nextflow consist of a series of processes. Processes specify data I/O and typically wrap around third-party software tools to process this data. Processes are connected through channels – asynchronous FIFO queues – which manage the flow of data throughout the pipeline.</p>
<p>Processes have the following basic structure:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">process</span> <span class="o">&lt;</span><span class="n">name</span><span class="o">&gt;</span> <span class="p">{</span>

    <span class="nb">input</span><span class="p">:</span>
    <span class="o">&lt;</span><span class="n">process</span> <span class="n">inputs</span><span class="o">&gt;</span>

    <span class="n">output</span><span class="p">:</span>
    <span class="o">&lt;</span><span class="n">process</span> <span class="n">outputs</span><span class="o">&gt;</span>

    <span class="n">script</span><span class="p">:</span>
    <span class="o">&lt;</span><span class="n">user</span> <span class="n">script</span> <span class="n">to</span> <span class="n">be</span> <span class="n">executed</span><span class="o">&gt;</span>
<span class="p">}</span>
</pre></div>
</div>
<p>Often, the script portion of the processes are reused by various sequencing pipelines. To help standardize pipeline development and ensure good practices are propogated to all pipelines, template processes are defined and inherited by pipeline processes.</p>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Templates are located in <strong>pipelines/templates</strong></p>
</div>
<p>For example, these two processes execute the same code:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1"># Without inheritance</span>
<span class="n">process</span> <span class="n">htseq</span> <span class="p">{</span>

    <span class="nb">input</span><span class="p">:</span>
    <span class="o">&lt;</span><span class="n">process</span> <span class="n">inputs</span><span class="o">&gt;</span>

    <span class="n">output</span><span class="p">:</span>
    <span class="o">&lt;</span><span class="n">process</span> <span class="n">outputs</span><span class="o">&gt;</span>

    <span class="n">script</span><span class="p">:</span>
    <span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    samtools view ${bamfiles} | htseq-count - ${gtf} \\</span>
<span class="sd">    --type   ${params.htseq.type}   \\</span>
<span class="sd">    --mode   ${params.htseq.mode}   \\</span>
<span class="sd">    --idattr ${params.htseq.idattr} \\</span>
<span class="sd">    --order  ${params.htseq.order}  \\</span>
<span class="sd">    &gt; counts.txt</span>
<span class="sd">    &#39;&#39;&#39;</span>
<span class="p">}</span>

<span class="c1"># With inheritance</span>
<span class="n">process</span> <span class="n">htseq</span> <span class="p">{</span>

    <span class="nb">input</span><span class="p">:</span>
    <span class="o">&lt;</span><span class="n">process</span> <span class="n">inputs</span><span class="o">&gt;</span>

    <span class="n">output</span><span class="p">:</span>
    <span class="o">&lt;</span><span class="n">process</span> <span class="n">outputs</span><span class="o">&gt;</span>

    <span class="n">script</span><span class="p">:</span>
    <span class="n">template</span> <span class="s1">&#39;htseq.sh&#39;</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<p>The RNA-seq pipeline output has the following basic structure:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>/pipeliner/RNA-seq
└── /results
    │
    ├── /sample_1
    │   ├── /trimgalore      | Trimmed Reads (.fq.gz) for sample_1
    │   ├── /fastqc
    │   ├── /rseqc
    │   └── /htseq
    │
    ├── /alignments          | Where (.bams) are saved
    ├── /aligner
    │   └── /index           | Index created and used during mapping
    │
    ├── /expression_matrix   | Aggregated count matrix
    ├── /expression_set      | An expression set (.rds) object
    ├── /reports             | Aggregated report across all samples pre/post pipeliner
    └── /logs                | Process-related logs
</pre></div>
</div>
<p>Each sample will have its own directory with sample-specific data and results for each process. Additionally, sequencing alignment files and the indexed reference genome will be saved for future use if specified. Summary reports pre/post-workflow can be found inside the reports directory.</p>
</div>
</div>
<span id="document-existing-pipelines"></span><div class="section" id="existing-pipelines">
<span id="existing-pipelines-page"></span><h2>Existing Pipelines<a class="headerlink" href="#existing-pipelines" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<span id="document-rna-seq"></span><div class="section" id="rna-seq">
<span id="rna-seq-page"></span><h3>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h3>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="check-reads-check-reads">
<h4>Check Reads <code class="code docutils literal notranslate"><span class="pre">check_reads</span></code><a class="headerlink" href="#check-reads-check-reads" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">None</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Ensures correct format of sequencing read files</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genome-indexing-hisat-indexing-star-indexing">
<h4>Genome Indexing <code class="code docutils literal notranslate"><span class="pre">hisat_indexing/star_indexing</span></code><a class="headerlink" href="#genome-indexing-hisat-indexing-star-indexing" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Genome reference file (<cite>.fa</cite>) | Genome annotation file (<cite>.gtf</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Directory containing indexed genome files</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses either <cite>STAR</cite> or <cite>HISAT2</cite> to build an indexed genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="pre-quality-check-pre-fastqc">
<h4>Pre-Quality Check <code class="code docutils literal notranslate"><span class="pre">pre_fastqc</span></code><a class="headerlink" href="#pre-quality-check-pre-fastqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Report files (<cite>.html</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>FastQC</cite> to check quality of read files</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="pre-multiqc-pre-multiqc">
<h4>Pre-MultiQC <code class="code docutils literal notranslate"><span class="pre">pre_multiqc</span></code><a class="headerlink" href="#pre-multiqc-pre-multiqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Log files (<cite>.log</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Summary report file (<cite>.html</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>MultiQC</cite> to generate a summary report</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="read-trimming-trim-galore">
<h4>Read Trimming <code class="code docutils literal notranslate"><span class="pre">trim_galore</span></code><a class="headerlink" href="#read-trimming-trim-galore" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Trimmed read files (<cite>.fastq</cite>) | Report files (<cite>.html</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Trims low quality reads with <cite>TrimGalore</cite> and checks quality with <cite>FastQC</cite></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="read-mapping-hisat-mapping-star-mapping">
<h4>Read Mapping <code class="code docutils literal notranslate"><span class="pre">hisat_mapping/star_mapping</span></code><a class="headerlink" href="#read-mapping-hisat-mapping-star-mapping" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>) | Genome annotation file (<cite>.gtf</cite>) | Directory containing indexed reference genome files</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">A list of alignment files (<cite>.bam</cite>) | Log files (<cite>.log</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses either <cite>STAR</cite> or <cite>HISAT2</cite> to align reads to a reference genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="reformat-reference-gtftobed">
<h4>Reformat Reference <code class="code docutils literal notranslate"><span class="pre">gtftobed</span></code><a class="headerlink" href="#reformat-reference-gtftobed" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Genome annotation file (<cite>.gtf</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Genome annotation file (<cite>.bed</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Converts genome annotation file from GTF to BED format</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="mapping-quality-rseqc">
<h4>Mapping Quality <code class="code docutils literal notranslate"><span class="pre">rseqc</span></code><a class="headerlink" href="#mapping-quality-rseqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">A list of alignment files (<cite>.bam</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Report files (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>RSeQC</cite> to check quality of alignment files</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="quantification-counting">
<h4>Quantification <code class="code docutils literal notranslate"><span class="pre">counting</span></code><a class="headerlink" href="#quantification-counting" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">A list of alignment files (<cite>.bam</cite>) | Genome annotation file (<cite>.gtf</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Read counts (<cite>.txt</cite>) | Log files (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses either <cite>StringTie</cite>, <cite>HTSeQ</cite>, or <cite>featureCounts</cite> to quantify reads</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="expression-matrix-expression-matrix">
<h4>Expression Matrix <code class="code docutils literal notranslate"><span class="pre">expression_matrix</span></code><a class="headerlink" href="#expression-matrix-expression-matrix" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">A list of count files (<cite>.txt</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">An expression matrix (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Reformats a list of count files into a genes x samples matrix</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="expression-features-expression-features">
<h4>Expression Features <code class="code docutils literal notranslate"><span class="pre">expression_features</span></code><a class="headerlink" href="#expression-features-expression-features" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Genome annotation file (<cite>.gtf</cite>) | An expression matrix (<cite>.txt</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Gene feature data (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Parses the genome annotation file for gene feature data</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="expression-set-expression-set">
<h4>Expression Set <code class="code docutils literal notranslate"><span class="pre">expression_set</span></code><a class="headerlink" href="#expression-set-expression-set" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">An expression matrix (<cite>.txt</cite>) | Gene feature data (<cite>.txt</cite>) | Sample phenotypic data (<cite>.txt</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">An expression set object (<cite>.rds</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Creates an expression set object with eData, fData, and pData</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="summary-report-multiqc">
<h4>Summary Report <code class="code docutils literal notranslate"><span class="pre">multiqc</span></code><a class="headerlink" href="#summary-report-multiqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Log files and summary reports from all processes</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">A summary report (<cite>.html</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>MultiQC</cite> to generate a summary report</td>
</tr>
</tbody>
</table>
</div>
</div>
<span id="document-scrna-seq"></span><div class="section" id="scrna-seq">
<span id="scrna-seq-page"></span><h3>scRNA-seq<a class="headerlink" href="#scrna-seq" title="Permalink to this headline">¶</a></h3>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="check-reads-check-reads">
<h4>Check Reads <code class="code docutils literal notranslate"><span class="pre">check_reads</span></code><a class="headerlink" href="#check-reads-check-reads" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">None</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Ensures correct format of sequencing read files</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="genome-indexing-hisat-indexing-star-indexing">
<h4>Genome Indexing <code class="code docutils literal notranslate"><span class="pre">hisat_indexing/star_indexing</span></code><a class="headerlink" href="#genome-indexing-hisat-indexing-star-indexing" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Genome reference file (<cite>.fa</cite>) | Genome annotation file (<cite>.gtf</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Directory containing indexed genome files</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses either <cite>STAR</cite> or <cite>HISAT2</cite> to build an indexed genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="quality-check-fastqc">
<h4>Quality Check <code class="code docutils literal notranslate"><span class="pre">fastqc</span></code><a class="headerlink" href="#quality-check-fastqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Report files (<cite>.html</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>FastQC</cite> to check quality of read files</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="whitelist-whitelist">
<h4>Whitelist <code class="code docutils literal notranslate"><span class="pre">whitelist</span></code><a class="headerlink" href="#whitelist-whitelist" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">A table of white listed barcodes (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>UMI-tools</cite> to extract and identify true cell barcodes</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="extract-extract">
<h4>Extract <code class="code docutils literal notranslate"><span class="pre">extract</span></code><a class="headerlink" href="#extract-extract" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>) | A table of white listed barcodes (<cite>.txt</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Extracted read files (<cite>.fastq</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>UMI-tools</cite> to extract barcode from reads and append to read name</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="read-mapping-hisat-mapping-star-mapping">
<h4>Read Mapping <code class="code docutils literal notranslate"><span class="pre">hisat_mapping/star_mapping</span></code><a class="headerlink" href="#read-mapping-hisat-mapping-star-mapping" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">List of read files (<cite>.fastq</cite>) | Genome annotation file (<cite>.gtf</cite>) | Directory containing indexed reference genome files</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">A list of alignment files (<cite>.bam</cite>) | Log files (<cite>.log</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses either <cite>STAR</cite> or <cite>HISAT2</cite> to align reads to a reference genome</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="reformat-reference-gtftobed">
<h4>Reformat Reference <code class="code docutils literal notranslate"><span class="pre">gtftobed</span></code><a class="headerlink" href="#reformat-reference-gtftobed" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Genome annotation file (<cite>.gtf</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Genome annotation file (<cite>.bed</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Converts genome annotation file from GTF to BED format</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="mapping-quality-rseqc">
<h4>Mapping Quality <code class="code docutils literal notranslate"><span class="pre">rseqc</span></code><a class="headerlink" href="#mapping-quality-rseqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">A list of alignment files (<cite>.bam</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Report files (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>RSeQC</cite> to check quality of alignment files</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="quantification-counting">
<h4>Quantification <code class="code docutils literal notranslate"><span class="pre">counting</span></code><a class="headerlink" href="#quantification-counting" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">A list of alignment files (<cite>.bam</cite>) | Genome annotation file (<cite>.gtf</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Read counts (<cite>.txt</cite>) | Log files (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>featureCounts</cite> to quantify reads</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="summary-report-multiqc">
<h4>Summary Report <code class="code docutils literal notranslate"><span class="pre">multiqc</span></code><a class="headerlink" href="#summary-report-multiqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Log files and summary reports from all processes</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">A summary report (<cite>.html</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>MultiQC</cite> to generate a summary report</td>
</tr>
</tbody>
</table>
</div>
</div>
<span id="document-dge"></span><div class="section" id="dge">
<span id="dege-page"></span><h3>DGE<a class="headerlink" href="#dge" title="Permalink to this headline">¶</a></h3>
<div class="toctree-wrapper compound">
</div>
<div class="section" id="quantification-counting">
<h4>Quantification <code class="code docutils literal notranslate"><span class="pre">counting</span></code><a class="headerlink" href="#quantification-counting" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">A list of alignment files (<cite>.bam</cite>) | Genome annotation file (<cite>.gtf</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">Read counts (<cite>.txt</cite>) | Log files (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>featureCounts</cite> to quantify reads</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="expression-matrix-expression-matrix">
<h4>Expression Matrix <code class="code docutils literal notranslate"><span class="pre">expression_matrix</span></code><a class="headerlink" href="#expression-matrix-expression-matrix" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">A list of count files (<cite>.txt</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">An expression matrix (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Reformats a list of count files into a genes x samples matrix</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="sample-renaming-rename-samples">
<h4>Sample Renaming <code class="code docutils literal notranslate"><span class="pre">rename_samples</span></code><a class="headerlink" href="#sample-renaming-rename-samples" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">An expression matrix (<cite>.txt</cite>)</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">An expression matrix (<cite>.txt</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Renames samples in expression matrix based on a user-supplied table</td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="summary-report-multiqc">
<h4>Summary Report <code class="code docutils literal notranslate"><span class="pre">multiqc</span></code><a class="headerlink" href="#summary-report-multiqc" title="Permalink to this headline">¶</a></h4>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">input:</th><td class="field-body">Log files and summary reports from all processes</td>
</tr>
<tr class="field-even field"><th class="field-name">output:</th><td class="field-body">A summary report (<cite>.html</cite>)</td>
</tr>
<tr class="field-odd field"><th class="field-name">script:</th><td class="field-body">Uses <cite>MultiQC</cite> to generate a summary report</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<span id="document-extending-pipelines"></span><div class="section" id="extending-pipelines">
<span id="extending-pipelines-page"></span><h2>Extending Pipelines<a class="headerlink" href="#extending-pipelines" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
</div>
</div>
</div>
</div>


           </div>
           
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2018, Anthony Federico, Stefano Monti.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1',
            LANGUAGE:'None',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>

  

  <script type="text/javascript" src="_static/js/theme.js"></script>

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>